File: contracts/api/clients/kotlin-spring-restclient/src/main/kotlin/app/epistola/client/EpistolaClient.kt
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.client

import app.epistola.client.auth.ApiKeyAuth
import app.epistola.client.auth.JwtSigner
import app.epistola.client.error.installProblemDetailHandler
import app.epistola.client.identity.ClientIdentity
import org.springframework.http.client.JdkClientHttpRequestFactory
import org.springframework.web.client.RestClient
import java.time.Duration

/**
 * The blessed setup: identity headers, JSON configuration, RFC 9457 problem parsing, and API-key
 * or self-signed-JWT authentication, assembled in one call instead of four.
 *
 * The pieces this assembles — [ClientIdentity], [ApiKeyAuth], [JwtSigner],
 * [epistolaMessageConverters], [installProblemDetailHandler] — stay independently usable and
 * independently documented; nothing here replaces them, and building a `RestClient` by hand from
 * them remains entirely supported. This exists because assembling them by hand is exactly where a
 * piece goes missing without anything failing loudly: wiring [epistolaMessageConverters] without
 * also calling [installProblemDetailHandler] compiles, runs, and every error response silently
 * comes back as a raw `RestClientResponseException` instead of a [app.epistola.client.error.ProblemDetailException] —
 * no exception, no log line, just a `catch` block that quietly stops matching. [builder] installs
 * both, every time, because there is no configuration in which a consumer of this entry point wants
 * only one of them.
 *
 * ```kotlin
 * val restClient = EpistolaClient.builder("https://epistola.example.com/api")
 *     .apiKey("epk_...")                              // or .jwtSigner(signer)
 *     .identity(ClientIdentity.builder().nodeId("pod-1").build())
 *     .build()
 *
 * val templatesApi = TemplatesApi(restClient)
 * ```
 *
 * ### Two timeout profiles from one builder
 *
 * A [Builder] holds the shared configuration — base URL, auth, identity, connect timeout — and
 * [Builder.build] can be called more than once, with a different [Builder.readTimeout] set in
 * between, for the two profiles a long-running consumer typically needs against the same backend:
 *
 * ```kotlin
 * val builder = EpistolaClient.builder(baseUrl).apiKey(apiKey).identity(identity)
 *
 * // Result-collector polling, preview rendering, catalog import: no read timeout at all, because
 * // a slow render or a large transfer is not a failure and should not be cut off mid-flight.
 * val longRunning = builder.readTimeout(null).build()
 *
 * // Everything else: bounded, so a wedged connection surfaces as an error instead of hanging.
 * val shortCalls = builder.readTimeout(Duration.ofSeconds(30)).build()
 * ```
 */
object EpistolaClient {

    /** A [Builder] with no base URL yet — call [Builder.baseUrl] before [Builder.build]. */
    fun builder(): Builder = Builder()

    /** A [Builder] with [baseUrl] already set. */
    fun builder(baseUrl: String): Builder = Builder().baseUrl(baseUrl)

    /** A [Builder] with [baseUrl] and a static [apiKey] already set — the common case in one call. */
    fun builder(baseUrl: String, apiKey: String): Builder = Builder().baseUrl(baseUrl).apiKey(apiKey)

    class Builder internal constructor() {
        private var baseUrl: String? = null
        private var identity: ClientIdentity? = null
        private var apiKey: String? = null
        private var jwtSigner: JwtSigner? = null
        private var connectTimeout: Duration = Duration.ofSeconds(10)
        private var readTimeout: Duration? = Duration.ofSeconds(30)

        fun baseUrl(baseUrl: String) = apply { this.baseUrl = baseUrl }

        /**
         * `User-Agent` + `X-EP-Node-Id`. Every request the contract accepts requires both; see
         * [ClientIdentity] for what happens if this is left unset.
         */
        fun identity(identity: ClientIdentity) = apply { this.identity = identity }

        /** `Authorization: ApiKey <key>`. Mutually exclusive with [jwtSigner] — whichever is called last wins. */
        fun apiKey(apiKey: String) = apply {
            this.apiKey = apiKey
            this.jwtSigner = null
        }

        /** `Authorization: Bearer <jwt>`, minted fresh per request. Mutually exclusive with [apiKey]. */
        fun jwtSigner(jwtSigner: JwtSigner) = apply {
            this.jwtSigner = jwtSigner
            this.apiKey = null
        }

        /** Time to establish the TCP connection (default 10s). Enforced in both timeout profiles. */
        fun connectTimeout(connectTimeout: Duration) = apply { this.connectTimeout = connectTimeout }

        /**
         * Time to wait for the response once the request is sent (default 30s). Pass `null` for no
         * read timeout at all — the JDK's own default, which is to wait indefinitely — for a
         * long-running operation (polling, rendering, a large transfer) that a fixed timeout would
         * cut off mid-flight rather than fail.
         */
        fun readTimeout(readTimeout: Duration?) = apply { this.readTimeout = readTimeout }

        /**
         * Builds a [RestClient] from the configuration so far. Safe to call more than once on the
         * same [Builder] — each call reads the builder's current state into an independent client,
         * which is what makes the two-timeout-profile pattern above work from one builder.
         */
        fun build(): RestClient {
            val url = requireNotNull(baseUrl) { "baseUrl is required" }

            // Not SimpleClientHttpRequestFactory: it wraps java.net.HttpURLConnection, which
            // rejects PATCH outright with "Invalid HTTP method: PATCH" — every one of the
            // contract's thirteen PATCH operations, including updateConsumer, would fail before
            // a request left the process. java.net.http.HttpClient has no such restriction.
            val httpClient = java.net.http.HttpClient.newBuilder()
                .connectTimeout(connectTimeout)
                .build()
            val requestFactory = JdkClientHttpRequestFactory(httpClient).apply {
                readTimeout?.let { setReadTimeout(it) }
            }

            val restClientBuilder = RestClient.builder()
                .baseUrl(url)
                .requestFactory(requestFactory)
                .epistolaMessageConverters()
                .installProblemDetailHandler()

            identity?.let { restClientBuilder.requestInterceptor(it.interceptor()) }
            val signer = jwtSigner
            val key = apiKey
            when {
                signer != null -> restClientBuilder.requestInterceptor(signer.interceptor())
                key != null -> restClientBuilder.requestInterceptor(ApiKeyAuth.of(key).interceptor())
            }

            return restClientBuilder.build()
        }
    }
}

File: contracts/api/clients/nodejs-fetch/src/http/epistolaClient.ts
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

import type { JwtSigner } from '../auth/jwtSigner.js'
import { Configuration, type FetchAPI, type Middleware } from '../generated/api/runtime.js'
import type { ClientIdentity } from '../identity/clientIdentity.js'
import { epistolaRequestMiddleware, problemDetailMiddleware } from './middleware.js'

type Auth = { readonly kind: 'api-key'; readonly key: string } | { readonly kind: 'jwt'; readonly signer: JwtSigner }

/** Everything an {@link EpistolaClient} is built from. */
export interface EpistolaClientOptions {
  /** The API base URL, including the contract's `/api` base path, e.g. `https://epistola.example.com/api`. */
  readonly baseUrl: string
  /** `User-Agent` + `X-EP-Node-Id`. Every request the contract accepts requires both. */
  readonly identity?: ClientIdentity
  /** `Authorization: ApiKey <key>`. Mutually exclusive with `jwtSigner`. */
  readonly apiKey?: string
  /** `Authorization: Bearer <jwt>`, minted fresh per request. Mutually exclusive with `apiKey`. */
  readonly jwtSigner?: JwtSigner
  /** The `fetch` to use; defaults to the platform's. */
  readonly fetchApi?: FetchAPI
  /** Aborts a request that has not completed within this many milliseconds. Default: no limit. */
  readonly requestTimeoutMs?: number
  /** Extra middleware, run after the Epistola request middleware and before the problem-detail one. */
  readonly middleware?: readonly Middleware[]
}

/**
 * The blessed setup: identity headers, API-key or self-signed-JWT authentication, the `Accept`
 * header each operation is declared with, and RFC 9457 problem parsing — assembled into the
 * `Configuration` every generated API class takes, so that
 *
 * ```ts
 * const client = EpistolaClient.builder('https://epistola.example.com/api', 'epk_...')
 *   .identity(ClientIdentity.builder().nodeId('pod-1').build())
 *   .build()
 *
 * const templates = new TemplatesApi(client)
 * ```
 *
 * is the whole of it. Problem parsing is not opt-in here, because forgetting it fails silently:
 * every error response would come back as a bare `ResponseError` with the problem document
 * unread, and nothing would say so.
 *
 * It is also what {@link ResultCollector} polls with: {@link requestHeaders} hands out the same
 * identity and a fresh token for the one endpoint the generated classes do not cover.
 */
export class EpistolaClient extends Configuration {
  /** The API base URL, without a trailing slash. */
  readonly baseUrl: string

  /** The identity headers this client sends, if configured. */
  readonly identity: ClientIdentity | undefined

  /** The `fetch` this client uses, for requests made outside the generated API classes. */
  readonly fetch: FetchAPI

  private readonly auth: Auth | undefined

  constructor(options: EpistolaClientOptions) {
    const baseUrl = normalizeBaseUrl(options.baseUrl)
    if (options.apiKey !== undefined && options.jwtSigner !== undefined) {
      throw new RangeError('apiKey and jwtSigner are mutually exclusive')
    }
    const auth: Auth | undefined = options.jwtSigner
      ? { kind: 'jwt', signer: options.jwtSigner }
      : options.apiKey !== undefined
        ? { kind: 'api-key', key: normalizeApiKey(options.apiKey) }
        : undefined
    const fetchApi = options.fetchApi ?? fetch
    // `authorization` is looked up through the instance so a fresh token is minted per request.
    let self: EpistolaClient | undefined
    super({
      basePath: baseUrl,
      fetchApi,
      middleware: [
        epistolaRequestMiddleware({
          basePath: baseUrl,
          identity: options.identity,
          authorization: () => self?.authorizationHeader(),
          requestTimeoutMs: options.requestTimeoutMs,
        }),
        ...(options.middleware ?? []),
        problemDetailMiddleware(),
      ],
    })
    self = this
    this.baseUrl = baseUrl
    this.identity = options.identity
    this.fetch = fetchApi
    this.auth = auth
  }

  /** A {@link EpistolaClientBuilder}, with `baseUrl` and a static `apiKey` already set when given. */
  static builder(baseUrl?: string, apiKey?: string): EpistolaClientBuilder {
    const builder = new EpistolaClientBuilder()
    if (baseUrl !== undefined) builder.baseUrl(baseUrl)
    if (apiKey !== undefined) builder.apiKey(apiKey)
    return builder
  }

  /**
   * The `Authorization` header value for a request made now — `ApiKey <key>`, or `Bearer <jwt>`
   * with a token minted for this call — or undefined when the client is unauthenticated.
   */
  authorizationHeader(): string | undefined {
    switch (this.auth?.kind) {
      case 'api-key':
        return `ApiKey ${this.auth.key}`
      case 'jwt':
        return `Bearer ${this.auth.signer.createToken()}`
      default:
        return undefined
    }
  }

  /**
   * The per-request Epistola headers — identity plus a fresh `Authorization` — for a request made
   * outside the generated API classes. {@link ResultCollector} drives the raw NDJSON collect
   * endpoint with these.
   */
  requestHeaders(): Record<string, string> {
    const headers: Record<string, string> = { ...(this.identity?.headers() ?? {}) }
    const authorization = this.authorizationHeader()
    if (authorization !== undefined) {
      headers.Authorization = authorization
    }
    return headers
  }
}

/** Fluent builder for {@link EpistolaClient}. */
export class EpistolaClientBuilder {
  private baseUrlValue: string | undefined
  private identityValue: ClientIdentity | undefined
  private apiKeyValue: string | undefined
  private jwtSignerValue: JwtSigner | undefined
  private fetchApiValue: FetchAPI | undefined
  private requestTimeoutMsValue: number | undefined
  private readonly middlewareValue: Middleware[] = []

  /** The API base URL, including the contract's `/api` base path. */
  baseUrl(baseUrl: string): this {
    this.baseUrlValue = baseUrl
    return this
  }

  /** `User-Agent` + `X-EP-Node-Id`. */
  identity(identity: ClientIdentity): this {
    this.identityValue = identity
    return this
  }

  /** `Authorization: ApiKey <key>`. Mutually exclusive with {@link jwtSigner} — whichever is called last wins. */
  apiKey(apiKey: string): this {
    this.apiKeyValue = normalizeApiKey(apiKey)
    this.jwtSignerValue = undefined
    return this
  }

  /** `Authorization: Bearer <jwt>`, minted fresh per request. Mutually exclusive with {@link apiKey}. */
  jwtSigner(signer: JwtSigner): this {
    this.jwtSignerValue = signer
    this.apiKeyValue = undefined
    return this
  }

  /** The `fetch` to use (advanced: a custom agent, proxies, recording). Defaults to the platform's. */
  fetchApi(fetchApi: FetchAPI): this {
    this.fetchApiValue = fetchApi
    return this
  }

  /**
   * Aborts a request that has not completed within this many milliseconds; pass undefined for no
   * limit, which is the default and the right setting for polling, rendering and large transfers.
   */
  requestTimeoutMs(timeoutMs: number | undefined): this {
    if (timeoutMs !== undefined && !(timeoutMs > 0)) {
      throw new RangeError('requestTimeoutMs must be positive')
    }
    this.requestTimeoutMsValue = timeoutMs
    return this
  }

  /** Adds middleware, run after the Epistola request middleware and before the problem-detail one. */
  middleware(...middleware: Middleware[]): this {
    this.middlewareValue.push(...middleware)
    return this
  }

  /**
   * Builds an {@link EpistolaClient} from the configuration so far. Safe to call more than once on
   * the same builder — each call reads the builder's current state into an independent client,
   * which is how one builder yields the two timeout profiles a long-running consumer needs.
   */
  build(): EpistolaClient {
    if (this.baseUrlValue === undefined) {
      throw new RangeError('baseUrl is required')
    }
    return new EpistolaClient({
      baseUrl: this.baseUrlValue,
      identity: this.identityValue,
      apiKey: this.apiKeyValue,
      jwtSigner: this.jwtSignerValue,
      fetchApi: this.fetchApiValue,
      requestTimeoutMs: this.requestTimeoutMsValue,
      middleware: [...this.middlewareValue],
    })
  }
}

function normalizeBaseUrl(baseUrl: string): string {
  let parsed: URL
  try {
    parsed = new URL(baseUrl)
  } catch (cause) {
    throw new RangeError(`baseUrl must be an absolute URL, got ${JSON.stringify(baseUrl)}`, { cause })
  }
  if (parsed.search !== '' || parsed.hash !== '') {
    throw new RangeError('baseUrl must not carry a query string or fragment')
  }
  return parsed.toString().replace(/\/+$/, '')
}

function normalizeApiKey(apiKey: string): string {
  const value = apiKey.trim()
  if (value === '') {
    throw new RangeError('apiKey must not be blank')
  }
  return value
}
